compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -173,21 +170,11 @@ pub(super) fn op_to_const<'tcx>(
             Immediate::ScalarPair(a, b) => {
                 debug!("ScalarPair(a: {:?}, b: {:?})", a, b);
                 // We know `offset` is relative to the allocation, so we can use `into_parts`.
-                let (data, start) = match a.to_pointer(ecx).unwrap().into_parts() {
-                    (Some(alloc_id), offset) => {
-                        (ecx.tcx.global_alloc(alloc_id).unwrap_memory(), offset.bytes())
-                    }
-                    (None, _offset) => (
-                        ecx.tcx.mk_const_alloc(Allocation::from_bytes_byte_aligned_immutable(
-                            b"" as &[u8],
-                        )),
-                        0,
-                    ),
-                };
+                let (alloc_id, start) = a.to_pointer(ecx).unwrap().into_parts();
                 let len = b.to_target_usize(ecx).unwrap();
-                let start = start.try_into().unwrap();
+                let start = start.bytes().try_into().unwrap();
                 let len: usize = len.try_into().unwrap();
-                ConstValue::Slice { data, start, end: start + len }
+                ConstValue::Slice { alloc_id, start, end: start + len }
```

Comment:
Uh, I just realized that valtree-to-const-value will hit this code path for all wide pointers, e.g. `&[u32]`. That doesn't seem right? Certainly we didn't write the `start`/`len` logic thinking about anything besides `u8`/`str`...

Reply:
We should probably generalize `ConstValue::Slice` to be used for arbitrary types that have a slice as their unsized tail. That justifies how valtrees are using them in valtree-to-constval conversion.

But that's for a future PR.

Reply:
I'm still hoping to be able to revive my ConstValue::Slice removal PR

Reply:
Honestly, with the simplification of op_to_const in this PR, the Slice variant isn't so bad any more. In particular these days it seems to be okay to have slices in `ConstValue::ByRef`, they don't *have to* be in `ConstValue::Slice`, and then it's pretty okay.

Reply:
Let's leave fixing the valtree problem to the next PR; this one is ready now.